msal/broker.py

```diff
@@ -0,0 +1,198 @@
+"""This module is an adaptor to the underlying broker.
+PyMsalRuntime is the package providing broker's functionality.
+
+PyMsalRuntime currently supports these Windows versions, listed in this MSFT internal link
+https://github.com/AzureAD/microsoft-authentication-library-for-cpp/pull/2406/files
+"""
+from threading import Event
+import json
+import logging
+import uuid
+
+import pymsalruntime  # ImportError would be raised on unsupported platforms such as Windows 8
+    # Its API description is available in site-packages/pymsalruntime/PyMsalRuntime.pyi
+
+logger = logging.getLogger(__name__)
+
+
+class RedirectUriError(ValueError):
+    pass
+
+
+class _CallbackData:
+    def __init__(self):
+        self.signal = Event()
+        self.auth_result = None
+
+    def complete(self, auth_result):
+        self.signal.set()
+        self.auth_result = auth_result
+
+
+def _convert_error(error, client_id):
+    context = error.get_context()  # Available since pymsalruntime 0.0.4
+    if "AADSTS50011" in context:  # In WAM, this could happen on both interactive and silent flows
+        raise RedirectUriError(  # This would be seen by either the app developer or end user
+            "MsalRuntime won't work unless this one more redirect_uri is registered to current app: "
+            "ms-appx-web://Microsoft.AAD.BrokerPlugin/{}".format(client_id))
+    return {
+        "error": "broker_error",  # Note: Broker implies your device needs to be compliant.
+            # You may use "dsregcmd /status" to check your device state
+            # https://docs.microsoft.com/en-us/azure/active-directory/devices/troubleshoot-device-dsregcmd
+        "error_description": "{}. Status: {}, Error code: {}, Tag: {}".format(
+            context,
+            error.get_status(), error.get_error_code(), error.get_tag()),
+        }
+
+
+def _read_account_by_id(account_id):
+    """Return the callback result which contains the account or error"""
+    callback_data = _CallbackData()
+    pymsalruntime.read_account_by_id(
+        account_id,
+        "correlation_id",
+        lambda result, callback_data=callback_data: callback_data.complete(result)
+        )
+    callback_data.signal.wait()
+    return callback_data.auth_result
+
+
+def _convert_result(result, client_id):  # Mimic an on-the-wire response from AAD
+    error = result.get_error()
+    if error:
+        return _convert_error(error, client_id)
+    id_token_claims = json.loads(result.get_id_token()) if result.get_id_token() else {}
+    account = result.get_account()
+    assert account, "Account is expected to be always available"
+    ## Note: As of pymsalruntime 0.1.0, only wam_account_ids property is available
+    #account.get_account_property("wam_account_ids")
```

Comment:
Is this still needed?

Reply:
Not strictly necessary but, if you still remember, back then it took you and I quite a while to reverse-engineer the available account properties. Leaving this inline comment here is my way to keep track of this piece of potentially useful information at the place it would likely be used. #Resolve